quarkchain/cluster/tests/test_jsonrpc.py

```diff
@@ -1190,26 +1141,74 @@ def test_createTransactions(self):
 
             send_request("createTransactions", {"numTxPerShard": 1, "xShardPercent": 0})
 
-    def test_subscribe(self):
+
+# ------------------------------- Test for JSONRPCWebsocketServer -------------------------------
+@contextmanager
+def jrpc_websocket_server_context(slave_server):
+    env = DEFAULT_ENV.copy()
+    env.cluster_config = ClusterConfig()
+    env.cluster_config.JSON_RPC_PORT = 38391
+    # to pass the circleCi
+    env.cluster_config.JSON_RPC_HOST = "127.0.0.1"
+
+    slave_config = SlaveConfig()
+    slave_config.HOST = "0.0.0.0"
+    slave_config.PORT = 38001
+    slave_config.WEBSOCKET_JSON_RPC_PORT = 38591
+    slave_config.ID = "S1"
+    slave_config.CHAIN_MASK_LIST = [5]
+
+    env.cluster_config.SLAVE_LIST.append(slave_config)
+    env.slave_config = env.cluster_config.get_slave_config("S1")
+    server = JSONRPCWebsocketServer.start_websocket_server(env, slave_server)
+    try:
+        yield server
+    finally:
+        server.shutdown()
+
+
+def send_websocket_request(request, num_response=1):
+    responses = []
+
+    async def __send_request(request):
+        uri = "ws://0.0.0.0:38591"
+        async with websockets.connect(uri) as websocket:
+            await websocket.send(request)
+            while True:
+                response = await websocket.recv()
+                responses.append(response)
+                if len(responses) == num_response:
+                    return responses
+
+    return call_async(__send_request(request))
+
+
+def send_websocket_ping(request):
+    async def __send_request(request):
+        uri = "ws://0.0.0.0:38591"
+        async with websockets.connect(uri) as websocket:
+            response = await WebSocketsClient(websocket).send(request)
+            return response
+
+    return call_async(__send_request(request))
+
+
+class TestJSONRPCWebsocket(unittest.TestCase):
+    def test_newHeads(self):
         id1 = Identity.create_random_identity()
         acc1 = Address.create_from_identity(id1, full_shard_key=0)
 
         with ClusterContext(
             1, acc1, small_coinbase=True
-        ) as clusters, jrpc_server_context(
+        ) as clusters, jrpc_http_server_context(
```

Comment:
`jrpc_websocket_server_context`?

Reply:
Slave server is connected through websocket while master server is connected through http in this test, which is the reason of using `jrpc_http_server_context` here.

Reply:
looks like master server doesn't need to start.